File: src/main/java/io/awilson/timein/domain/Student.java
package io.awilson.timein.domain;

import lombok.*;

import javax.persistence.*;
import java.time.Duration;
import java.util.List;

@Data
@Entity
@ToString(exclude = {"instructor", "sessions"})
@EqualsAndHashCode(exclude = {"instructor"})
public class Student {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Integer id;
    @Version
    private Integer version;
    private String firstName;
    private String lastName;
    @ManyToOne
    private Instructor instructor;
    @ManyToOne
    private Course course;
    @OneToMany(mappedBy = "student")
    private List<Session> sessions;
    @OneToOne
    private Session currentSession;

    /**
     * Does the student have any active sessions?
     * @return
     */
    public Boolean anyActive() {
        return sessions.stream()
                .anyMatch(x -> x.getTimeEnd() == null);
    }

    public String getFullName() {
        return firstName + " " + this.lastName;
    }

    public Duration getTime() {
        return sessions.stream()
                .map(Session::getDuration)
                .reduce((x, y) -> x.plus(y))
                .get();
    }
}

File: src/main/java/io/awilson/timein/domain/Student.kt
package io.awilson.timein.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import javax.persistence.*

@Entity
class Student {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    var id: Int = 0
    @Version
    var version: Int = 0
    var firstName: String = ""
    var lastName: String = ""
    @JsonIgnore
    @ManyToOne
    var instructor: Instructor? = null
    @JsonIgnore
    @ManyToOne
    var course: Course? = null
    @JsonIgnore
    @OneToMany(mappedBy = "student")
    var sessions: MutableList<Session> = ArrayList()
    @JsonIgnore
    @OneToOne
    var currentSession: Session? = null
    var active: Boolean = true

    /**
     * Does the student have any active sessions?

     * @return true if the student has at least one active session.
     */
    fun anyActive(): Boolean = sessions.stream().anyMatch { it.isOpen }

    val instructorId: Int?
        get() = instructor?.id

    val currentSessionId: Int?
        get() = currentSession?.id

    val courseId: Int?
        get() = course?.id

    val isOnline: Boolean
        get() = !(currentSession == null)

    val fullName: String
        get() = "$firstName $lastName"
}
